### PR TITLE
Fix edge cases for deploy cycle breakdown

### PR DESCRIPTION
**Pull request by snabar**
Handle cases and add tests for when:
1. Deployments have no associated builds
2. Builds have no associated commits

### CODE REVIEW

File: apps/dora-metrics/deploy-cycle-breakdown/app.py [hunk 1]
```diff
@@ -10,41 +10,49 @@ class DeployStages:
...
-                int(pr["mergedAt"]) - int(first_approval["submittedAt"])
-        else:
-            self.pr_merge_time = \
-                int(pr["mergedAt"]) - int(pr["createdAt"]) if pr else None
-            self.pr_review_time = None
-        self.build_time = int(deployment["build"]["endedAt"]) - \
-            int(deployment["build"]["startedAt"])
-        self.deploy_time = \
-            int(deployment["endedAt"]) - int(deployment["startedAt"])
+        try:
```
Comment (snabar) [inline apps/dora-metrics/deploy-cycle-breakdown/app.py:22]: I added a try-except here in case any of the entities below don't have any of the fields we're referencing, e.g. if the build doesn't have an "endedAt" field. Not sure what the behavior in the "except" should be? Is it okay to just print? @mtth

Comment (mtth) [reply apps/dora-metrics/deploy-cycle-breakdown/app.py:22]: Printing is not nice to users, it can't be turned off. I'd either introduce a logger or add a `TODO` to do so and do nothing for now. It's more an expected case than an error, no?

Also, w.r.t. catching `KeyError`:

+ You can omit the `if ... in ...` guards above since they are now redundant.
+ Won't the timestamp be present but `None` in that case, meaning we'd get a different error?

Comment (snabar) [reply apps/dora-metrics/deploy-cycle-breakdown/app.py:22]: The if ... in ... guards are not quite redundant, because we still try to compute the rest of what can be computed. E.g. if there's no builds, we still compute the deploy time. If there's no commits, we still try to compute the build time and the deploy time. If there's no pr approvals, we still try to compute the merge time time, build time, deploy time etc.

Comment (snabar) [reply apps/dora-metrics/deploy-cycle-breakdown/app.py:22]: @mtth I added in another kind of error to catch -- a TypeError in case the int() casts fail because the timestamps are None. But it would be nice if I could generally just catch any error. Flake8 complains if I just put in an all-encompassing "except:" though.

File: apps/dora-metrics/deploy-cycle-breakdown/app.py [hunk 2]
```diff
@@ -51,8 +51,9 @@ def __init__(self, deployment):
                     int(deployment["build"]["startedAt"])
             self.deploy_time = \
                 int(deployment["endedAt"]) - int(deployment["startedAt"])
-        except KeyError:
-            print("Cannot compute deploy stages for deployment %s" % self.id)
+        except TypeError or KeyError:
```
Comment (mtth) [inline apps/dora-metrics/deploy-cycle-breakdown/app.py:54]: Can you double-check the syntax? I haven't seen this before.

https://stackoverflow.com/questions/6470428/catch-multiple-exceptions-in-one-line-except-block

File: apps/dora-metrics/deploy-cycle-breakdown/app.py (hunk 1 above)
Comment (mtth) [reply apps/dora-metrics/deploy-cycle-breakdown/app.py:22]: More specific errors are better to avoid hiding bugs (particularly important here since we're swallowing it). Given the growing errors caught and the size of the `try` block, I'd now recommend handling these cases with explicit guards. It's more code upfront but will be much easier to debug going forward, for example if something later changes in the data.

FTR you can also suppress errors. https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html

File: apps/dora-metrics/deploy-cycle-breakdown/app.py [hunk 3]
```diff
@@ -10,41 +11,61 @@ class DeployStages:
...
-                int(pr["mergedAt"]) - int(first_approval["submittedAt"])
-        else:
-            self.pr_merge_time = \
-                int(pr["mergedAt"]) - int(pr["createdAt"]) if pr else None
-            self.pr_review_time = None
-        self.build_time = int(deployment["build"]["endedAt"]) - \
-            int(deployment["build"]["startedAt"])
-        self.deploy_time = \
-            int(deployment["endedAt"]) - int(deployment["startedAt"])
+        if "build" in deployment and deployment["build"] is not None:
```
Comment (chalenge) [inline apps/dora-metrics/deploy-cycle-breakdown/app.py:23]: how about negating this to reduce nesting. Something like
```python
if ((not "build" in deployment and deployment["build"] is None) || (not deployment["build"]["commitAssociations"]):
return

commits = [
      node["commit"]
      for node in deployment["build"]["commitAssociations"]["nodes"]
]
```
or maybe 
```python
commitAssoc = getattr(getattr(deployment, 'build', None), 'commitAssociations', None)
if not commitAssoc:
  return 
```

File: apps/dora-metrics/deploy-cycle-breakdown/app.py [hunk 4]
```diff
@@ -10,41 +11,61 @@ class DeployStages:
...
+                    for node in deployment["build"]["commitAssociations"]["nodes"]
+                ]
+                if commits:
+                    commits.sort(key=lambda c: c["createdAt"])
+                    last_commit = commits[-1]
+                    pr = last_commit["mergedPullRequest"]
+                    approvals = [
+                        r for r in pr["reviews"]["nodes"]
+                        if r["state"] == "approved" and r["submittedAt"]
+                    ] if pr else None
```
Comment (chalenge) [inline apps/dora-metrics/deploy-cycle-breakdown/app.py:36]: formatting makes it confusing to read

File: apps/dora-metrics/deploy-cycle-breakdown/app.py [hunk 5]
```diff
@@ -10,41 +11,61 @@ class DeployStages:
...
+                    if approvals:
+                        approvals.sort(key=lambda a: a["submittedAt"])
+                        first_approval = approvals[0]
+                        self.pr_review_time = self.diff(
```
Comment (chalenge) [inline apps/dora-metrics/deploy-cycle-breakdown/app.py:40]: can't we use the diff utils in the client

File: apps/dora-metrics/deploy-cycle-breakdown/app.py [hunk 6]
```diff
@@ -10,41 +11,61 @@ class DeployStages:
...
+                            first_approval["submittedAt"], pr["createdAt"],
+                            'pr_review_time'
+                        )
+                        self.pr_merge_time = self.diff(
+                            pr["mergedAt"], first_approval["submittedAt"],
+                            'pr_merge_time'
+                        )
+                    else:
+                        self.pr_merge_time = self.diff(
```
Comment (chalenge) [inline apps/dora-metrics/deploy-cycle-breakdown/app.py:49]: With the if on the last like maybe hard to follow. I might put a linter ignore rule to make it readable

File: apps/dora-metrics/deploy-cycle-breakdown/app.py (hunk 5 above)
Comment (snabar) [reply apps/dora-metrics/deploy-cycle-breakdown/app.py:40]: this is just attempting to subtract two integers and logging an error if it cannot

File: apps/dora-metrics/deploy-cycle-breakdown/app.py (hunk 4 above)
Comment (snabar) [reply apps/dora-metrics/deploy-cycle-breakdown/app.py:36]: what's a better formatting?

**Review by chalenge — APPROVED** [review]
